Address Comments from Demo 1

gabrielleungkh opened the pull request:
- Changed error checking within the builder classes so that they don't throw exceptions, but are handled gracefully so no crashes occur
- Made ConcreteMediator and CarrierManagementSystemDirector singletons have synchronized getInstance methods
- In UI, instead of having enum class for defining the number of RF ports for LTE/WCDMA, changed to static final constants
- Changed concrete FrequencyBands and RfPort enums to only have the frequency/port letter respectively

- To update CarrierManagementSystemClient with updated expected test case results

gabrielleungkh committed:
Address Comments from Demo 1

- Changed error checking within the builder classes so that they don't throw exceptions, but are handled gracefully so no crashes occur
- Made ConcreteMediator and CarrierManagementSystemDirector singletons have synchronized getInstance methods
- In UI, instead of having enum class for defining the number of RF ports for LTE/WCDMA, changed to static final constants
- Changed concrete FrequencyBands and RfPort enums to only have the frequency/port letter respectively

- To update CarrierManagementSystemClient with updated expected test case results

john-breton commented on src/main/java/carriermanagementsystem/CarrierManagementSystemDirector.java:
```diff
@@ -19,48 +19,47 @@
-	// This list will save all carrier object and carrierID so the object can be
-	// modified from radio unit.
-	// private ConcurrentHashMap<Integer,Carrier> carrierList;
+    // For the future usage.
```
Since RadioUnits will have a list of Carriers, this comment here can probably be removed entirely.

gabrielleungkh replied:
Done

john-breton commented on src/main/java/carriermanagementsystem/CarrierManagementSystemDirector.java:
```diff
@@ -19,48 +19,47 @@
 public class CarrierManagementSystemDirector implements CarrierManagementIf {
-    private static volatile CarrierManagementSystemDirector UNIQUE_INSTANCE = new CarrierManagementSystemDirector();
-    private final int LTE_MAXIMUM_CARRIER_ID = 500; 
+    private static final CarrierManagementSystemDirector UNIQUE_INSTANCE = new CarrierManagementSystemDirector();
```
This should still be volatile (at least according to what Ista deems as the only "correct" Singleton). And the change should be in the getInstance() method.

gabrielleungkh replied:
I've been trying to figure out how volatile vs synchronized should be used and I saw most times it was either one or the other, but I'm not sure enough of that so I'll just put it back :)

john-breton commented on src/main/java/carriermanagementsystem/CarrierManagementSystemDirector.java:
```diff
@@ -19,48 +19,47 @@
      * @return CarrierManagementSystemDirector singleton instance
      */
-    public static CarrierManagementSystemDirector getInstance() {
+    public static synchronized CarrierManagementSystemDirector getInstance() {
```
This method should be changed to something like the following: 
```
public static CarrierManagementSystemDirector getInstance() {
    if (UNIQUE_INSTANCE == null) {
        synchronized (CarrierManagementSystemDirector.class) {
            if (UNIQUE_INSTANCE == null) {
                UNIQUE_INSTANCE = new CarrierManagementSystemDirector();
            }
        }
    }
    return UNIQUE_INSTANCE;
}
```

gabrielleungkh replied:
I was talking with Ramin over this part with the double-checked locking, and since we chose eager instantiation as a way to keep the entire thing thread safe, I think just making getInstance() synchronized should be enough?

john-breton commented on src/main/java/common/Carrier.java:
```diff
@@ -10,7 +10,7 @@
 public abstract class Carrier {
     protected final Integer carrierId;
-    protected final List<RfPort> rfPorts;
+    protected List<RfPort> rfPorts = null;
```
This doesn't need to be explicitly initialized to null here.

gabrielleungkh replied:
When we were unit testing, it errored out without the explicit initialization, but I just tried again and it's working fine without so I'll take this out. Must have been something weird when we were still testing our change.

john-breton commented on src/main/java/common/RatType.java:
```diff
@@ -5,8 +5,9 @@
-	WCDMA("WCDMA");
+	
+    LTE("Long-Term Evolution"),
+	WCDMA("Wideband Code Division Multiple Access");
```
When Ista assumed we all meant what WCDMA stood for I felt like I didn't belong as an Ericsson employee.

john-breton commented on src/main/java/mediator/ConcreteMediator.java:
```diff
@@ -49,7 +49,7 @@ private ConcreteMediator() {
      * @return The Singleton instance of the ConcreteMediator class.
      */
-    public static ConcreteMediator getInstance() {
+    public static synchronized ConcreteMediator getInstance() {
```
This will be getting updated in a separate commit. Basically exactly what I mentioned above for CarrierManagementSystemDirector.

gabrielleungkh replied:
In that case I'll take this out and leave it for the separate commit from you to come in. I'd still like to discuss whether we should go forward with double checked locking over eager instantiation at some point though.

john-breton reviewed (commented):
These changes look good. The only additional thing I remember Ista mentioning was about combining the enums into a single file, but that can be done at a later time/not at all if we don't agree with it.
Some minor comments for now but I'll approve it regardless since everything I mentioned would be easy fixes.

gabrielleungkh committed:
Address John's Comments + Update test case results

- Addressed John's comments
- Updated CarrierManagementSystemClient so the expected results for the forced failure test cases reflect the updated error handling logic

matthewkwsiu commented on src/main/java/carriermanagementsystem/CarrierManagementIf.java:
```diff
@@ -9,35 +10,34 @@
-     *                           LTE carrier frequency are available in
-     *                           CarrierFrequencies enum class.
-     * @param transmittingPower  LTE carrier need transmitting power to be created.
+     * @param rfPorts           LTE carrier needs 4 ports. These RF ports can be
```
Nice work on all the javadocs!

matthewkwsiu commented on src/main/java/carriermanagementsystem/CarrierManagementSystemClient.java:
```diff
@@ -50,16 +50,16 @@ public static void main(String[] args) {
-//         * LTE RF Ports: RF_3 = D  RF_4 = E  RF_5 = F  RF_6 = G  
+//         * LTE Carrier ID: 15
+//         * LTE Carrier Frequency: 1710 MHz
+//         * LTE RF Ports: D  E  F  G  
```
These were to address his points on just using letters right?

matthewkwsiu commented on src/main/java/carriermanagementsystem/CarrierManagementSystemClient.java:
```diff
@@ -68,27 +68,38 @@ public static void main(String[] args) {
 //         */
 //        
-//        FrequencyBand WcdmaBand_6 = FrequencyBand.WCDMA_BAND_6;
+//        FrequencyBand WcdmaBand_5 = FrequencyBand.WCDMA_BAND_5;
```
Pardon, to quickly clarify, was there a reason for changing these bands in response to a point of feedback Ista had gave? 

Or was this for testing?

gabrielleungkh replied:
I realized the variable names just didn't match up with the examples so I changed them back

matthewkwsiu commented on src/main/java/carriermanagementsystem/CarrierManagementSystemDirector.java:
```diff
@@ -20,47 +20,42 @@
 public class CarrierManagementSystemDirector implements CarrierManagementIf {
     private static volatile CarrierManagementSystemDirector UNIQUE_INSTANCE = new CarrierManagementSystemDirector();
-    private final int LTE_MAXIMUM_CARRIER_ID = 500; 
+    private final int LTE_MAXIMUM_CARRIER_ID = 500;
```
Nice whitespace changes. Was this from CTRL + SHIFT + F?

matthewkwsiu commented on src/main/java/carriermanagementsystem/WcdmaCarrierBuilder.java:
```diff
@@ -8,9 +8,11 @@
+ * objects.
  */
 public class WcdmaCarrierBuilder implements CarrierBuilder {
+    private static final int WCDMA_RF_PORTS_NUMBER = 2;
```
I appreciate you adding this variable to remove the magic number (just putting '2' in the code) from this class. Nice work.

matthewkwsiu commented on src/main/java/userinterface/UserInterface.java:
```diff
@@ -144,25 +146,25 @@ public static void main(String[] args) {
-        }
-
-    }
+//    /**
```
To check, are you keeping these as commented out since you'd like to keep them as reference, or should these be removed?

matthewkwsiu reviewed (commented):
Thanks for addressing the changes that Ista had done. I've left some comments and questions. Nice work!

One small note: I appreciate that you adjusted formatting and spacing in some of these files (I assume you used CTRL + SHIFT + F) but it can sometimes make it a little more challenging to review the commit because of how all the whitespace changes are treated as individual changes and thus crowds out some of the practical changes you had done. This was feedback I had received in the past. I'm not sure if applies very much here, especially as this is a casual project, but I thought I'd share the feedback with you.

john-breton reviewed (approved):
Everything looks good to me, I don't think we need to delay this getting merged any longer.

gabrielleungkh committed:
Addressing David's Comments and more code cleanup

- Added getRatType() method to Carrier
- Added a similar check in the CarrierBuilder classes to do error checking on the FrequencyBand passed in and fail gracefully
- Noticed that CarrierFrequencies was still used in the javadoc which should have been replaced with FrequencyBand so made those updates
- Changed the method name for getCarrierFrequencies() in Carrier to getFrequencyBand()

john-breton committed:
Merge branch 'main' into Demo1Comments
